### ISSUE TITLE
regression: cannot return value referencing local variable

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#82667]
https://crater-reports.s3.amazonaws.com/beta-1.51-1/beta-2021-02-14/gh/bow.gtetools/log.txt regressed in the beta crater run. Unfortunately the span points into a quick_error macro call, so hard to say exactly whether this is caused by something wrong in the compiler code or just changes to inference or something like that.

Probably next steps are to bisect and/or get a MCVE of that crate.

**Comment by SNCPlay42** [comment]
Reduced:

```rust
use std::error::Error;
use std::fmt;

#[derive(Debug)]
pub enum IntervalError {
    InvalidRange,
}
impl fmt::Display for IntervalError {
    fn fmt(&self, _fmt: &mut fmt::Formatter) -> fmt::Result {
        unimplemented!()
    }
}

impl Error for IntervalError {}

#[derive(Debug)]
pub enum ModelError {
    InvalidInterval(IntervalError),
}

impl ModelError {
    pub fn description(&self) -> &str {
        match *self {
            ModelError::InvalidInterval(ref err) => match err {
                ref otherwise => otherwise.description(),
            },
        }
    }
}
```

The type of `otherwise` is `&&IntervalError`. Prior to #75180, this was autoderefed; now, `&impl Error` implements `Error` so no autoderef happens and so the function returns a value that borrows `otherwise`.

Similar regressions were [observed](https://github.com/rust-lang/rust/pull/75180#issuecomment-681492119) in #75180's crater run, but this particular repository [failed to build on stable](https://crater-reports.s3.amazonaws.com/pr-75180/master%23119d2a1a98fe87d4ae6cabf12134a0ef2fb95851/gh/bow.gtetools/log.txt) in that crater run. Something similar happened to [another crate](https://github.com/rust-lang/rust/pull/75180#issuecomment-767632568), and the issue was [since fixed](https://github.com/rust-lang/crater/issues/560). It turned out to involve path dependencies, and this crate has one.

In short, this is #81460, which has been declared acceptable breakage.

Bisection points to #75180's rollup.

@rustbot label -E-needs-mcve -E-needs-bisection

**Comment by apiraino** [comment]
as per the issue analysis (thanks @SNCPlay42 !), removing the prioritize label

@rustbot label -I-prioritize

**Comment by JohnTitor** [comment]
Triage: Closing this as another instance of #81460, as mentioned above (https://github.com/rust-lang/rust/issues/82667#issuecomment-788281442).
